win32ss/user/ntuser/msgqueue.c

```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
         // to switch the keyboard layout and store it to global variable.
         // Using this combination of hotkeys in this function
 
-        if ( gdwLanguageToggleKey < 3 &&
-             IS_KEY_DOWN(gafAsyncKeyState, gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL) )
+        if ( gdwLanguageToggleKey < 3 )
         {
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            LanguageHotkey = gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL;
+            if( IS_KEY_DOWN(gafAsyncKeyState, LanguageHotkey) ||
+               IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT) )
             {
-                WPARAM wParamILR;
-                PKL pkl = pti->KeyboardLayout;
+               if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))) {
+                  bChangeLanguage = TRUE;
+               }
+               else {
+                  if ( wParam == VK_LMENU && wParam == LanguageHotkey
+                       && !(IS_KEY_DOWN(gafAsyncKeyState, VK_LMENU)) )
+                  {
+                     if ( IS_KEY_DOWN(gafAsyncKeyState, VK_LSHIFT) )
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_FORWARD;  // Left Alt - Left Shift, Next
+                     else if ( IS_KEY_DOWN(gafAsyncKeyState, VK_RSHIFT) )
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_BACKWARD; // Left Alt - Right Shift, Previous
+
+                     bChangeLanguage = TRUE;
+                  }
+                  else if ( Msg->wParam == VK_CONTROL && Msg->wParam == LanguageHotkey
+                           && !(IS_KEY_DOWN(gafAsyncKeyState, VK_CONTROL)) )
+                  {
+                     if ( IS_KEY_DOWN(gafAsyncKeyState, VK_LSHIFT) )
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_FORWARD;  // Left Alt - Left Shift, Next
```

Comment:
Aren't you looking at the Left/Right control key (+ the left/right shift) instead? This seems to change the keyboard layout only, within a fixed language.

Reply:
Language in most cases also changes the layout too. And hotkey is configurable. Actually ROS does not support layout change consistently, there are only parts implemented yet. So this is the nearest for JIRA ticket context. There are future todos but not in this context.